### DISCUSSION TITLE
MQTT hour price values not accepted by Domotics

### DISCUSSION THREAD

**Discussion by ksdale** [UtilitechAS/amsreader-firmware#873] [General]
I'm using MQTT with topic 'homeassistant' to feed data to Domoticz, and it works fine, except for hour prices.
This is the same for all hours up to hour 36 (hour 37 in the prices struct is not referred to).

I look into it and have noticed some differences in the JSON structs.

The "val_tpl" in topic homeassistant/sensor/Cheapest1hr looks like:
"{{ value_json.prices.cheapest1hr | is_defined }}"

compared to val_tpl in topic for homeassistant/sensor/ams-cf58_prices0:
"{{ value_json.prices['0'] | is_defined }}"

I noted the difference between the two val_tpl references:
   **value_json.prices.cheapest1hr**  and  **value_json.prices['0']**

cheapest1hr  is autodetected by Domoticz, prices['0'] is not. 

From the Domoticz error log:
2024-11-20 16:00:00.185 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[0]) 

The prices struct look like:
{
  "id": "58:CF:79:9B:4C:7C",
  "prices": {
    "0": 1.9741,
    "1": 1.7019,
 ...
    "36": null,
    "37": null,
    "min": 0.8459,
    "max": 1.9741,
    "cheapest1hr": "2024-11-22T22:00:00Z",
    "cheapest3hr": "2024-11-22T01:00:00Z",
    "cheapest6hr": "2024-11-21T23:00:00Z"
  }
}

So to me it looks as the correct reference should be **value.prices.0.**

**Comment by ArnieO** [comment]
Huh. We do not have many users on Domoticz - but many on Home Assistant using the 'homeassistant' payload - without any such issue as you describe.

I am wondering whether the issue you see is some kind of misinterpretation in Domoticz?
To compare: What do you see in val_tpl in topic for homeassistant/sensor/ams-cf58_prices1?

**Reply by ksdale** [reply]
homeassistant/sensor/ams-cf58_prices1 :  "val_tpl": "{{ value_json.prices['1'] | is_defined }}"
homeassistant/sensor/ams-cf58_prices2 :  "val_tpl": "{{ value_json.prices['2'] | is_defined }}"
 and so on.

**Reply by ArnieO** [reply]
> homeassistant/sensor/ams-cf58_prices1 : "val_tpl": "{{ value_json.prices['1'] | is_defined }}" homeassistant/sensor/ams-cf58_prices2 : "val_tpl": "{{ value_json.prices['2'] | is_defined }}" and so on.

So basically the same as for homeassistant/sensor/ams-cf58_prices0?
Could your issue be an error in how Domoticz reads/interprets the data?
Have you considered using one of the other payload formats? 
I know the payload for Domoticz is lacking a lot - but one of the other formats might be better suited?

**Comment by ksdale** [comment]
One more question:
Can you tell the difference (or equality) between:
  value_json.prices['0'] ,  value_json.prices[0] and value_json.prices.0

**Reply by ArnieO** [reply]
The right person to answer that is @gskjold (when he finds time).

**Comment by rogerthn2019** [comment]
Running domoticz Version: 2024.7
39 lines in domoticz log every hour as below, not critical but annoying

> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.exportprices[0])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[0])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[1])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[10])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[11])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[12])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[13])
> 2024-11-26 19:00:00.329 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[14])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[15])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[16])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[17])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[18])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[19])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[2])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[20])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[21])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[22])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[23])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[24])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[25])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[26])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[27])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[28])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[29])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[3])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[30])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[31])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[32])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[33])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[34])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[35])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[36])
> 2024-11-26 19:00:00.330 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[37])
> 2024-11-26 19:00:00.331 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[4])
> 2024-11-26 19:00:00.331 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[5])
> 2024-11-26 19:00:00.331 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[6])
> 2024-11-26 19:00:00.331 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[7])
> 2024-11-26 19:00:00.331 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[8])
> 2024-11-26 19:00:00.331 Error: MQTT: Exception (GetValueFromTemplate): in Json::Value::operator[](ArrayIndex): requires arrayValue! (Template: value_json.prices[9])
>
